### PR TITLE
Upgrade to rapidhash

### PR DESCRIPTION
**Pull request by Nicoshev**
Upgrading 64-bit from [wyhash](https://github.com/wangyi-fudan/wyhash) to [rapidhash](https://github.com/Nicoshev/rapidhash).

The wyhash owner has recognized rapidhash as the official successor in its repo.

rapidhash has shown to be faster and of higher quality than wyhash.

It is also officially compatible with MSVC across all architectures.

## References and Relevant Issues

rapidhash is currently the fastest recommended hash function by [SMHasher](https://github.com/rurban/smhasher?tab=readme-ov-file#summary).  

rapidhash is currently the fastest passing hash in [SMHasher3](https://gitlab.com/fwojcik/smhasher3/-/blob/main/results/README.md#passing-hashes).  

[Collision-based study](https://github.com/Nicoshev/rapidhash/tree/master?tab=readme-ov-file#collision-based-hash-quality-study) showing a collision probability lower than wyhash and close to ideal.

### CODE REVIEW

**Comment by Nicoshev** [comment]
@microsoft-github-policy-service agree

**Review by DHowett — COMMENTED** [review]
I'm gating on @lhecker as the final signoff, as he's the one who chose and integrated wyahsh

File: src/inc/til/rapidhash.h [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+/*
+ *  Reads and combines 3 bytes of input.
+ *
+ *  @param p  Buffer to read from.
+ *  @param k  Length of @p, in bytes.
+ *
+ *  Always reads and combines 3 bytes from memory.
+ *  Guarantees to read each buffer position at least once.
+ *  
+ *  Returns a 64-bit value contaning all three bytes read.
```
[the remark this answers is not in the document]
Comment (Nicoshev) [reply src/inc/til/rapidhash.h]: Thanks, fixed

File: src/inc/til/rapidhash.h [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+ *  64*64 -> 128bit multiply function.
+ *  
+ *  @param A  Address of 64-bit number.
+ *  @param B  Address of 64-bit number.
+ *
+ *  Calculates 128-bit C = *A * *B.
+ *
+ *  When RAPIDHASH_FAST is defined:
+ *  Overwritres A contents with C's low 64 bits.
+ *  Overwritres B contents with C's high 64 bits.
```
[the remark this answers is not in the document]
Comment (Nicoshev) [reply src/inc/til/rapidhash.h]: Thanks, fixed

File: src/inc/til/rapidhash.h [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+ *
+ *  @param p  Buffer to read from.
+ *  @param k  Length of @p, in bytes.
+ *
+ *  Always reads and combines 3 bytes from memory.
+ *  Guarantees to read each buffer position at least once.
+ *  
+ *  Returns a 64-bit value containing all three bytes read. 
+ */
+static inline uint64_t rapid_readSmall(const uint8_t* p, size_t k) RAPIDHASH_NOEXCEPT { return ((static_cast<uint64_t>(p[0])) << 56) | ((static_cast<uint64_t>(p[k >> 1])) << 32) | p[k - 1]; }
```
Comment (lhecker) [inline src/inc/til/rapidhash.h:268]: But if we do end up vendoring the `rapidhash.h` header I think it'd be best to keep it entirely unmodified. If the "AuditMode" compilation complains or any other part of the CI, I'd rather disable the checks.

**Review by lhecker — APPROVED** [review]

File: src/inc/til/rapidhash.h (hunk 3 above)
Comment (Nicoshev) [reply src/inc/til/rapidhash.h]: Talked offline, decided to move header file to oss folder

**Comment by DHowett** [comment]
Thanks for all the effort here, Nicolas!

I have some concerns, plus a couple "favorite things" about what we're doing today.

Right now, our implementation of wyhash is fully contained and inlined into `til`, and nobody needs to include any further headers or take on any `static`s[^1].

I'm concerned that we are also only replacing the 64-bit hash function, so we're not fully pulling wyhash from our code. We will forever have _two_ OSS hash library dependencies rather than just the one.

All that said, I'd still be interested... but we are also not using it for any data larger than, say, 1kb.

I'm not certain the performance boost for our specific workload is worth switching, all other things taken into consideration. If we were going to be processing untold heaps of data I might feel differently, but... eh. It's written now, but so is the code we had before. 🤷🏻

[^1]: Doesn't the compiler emit one of [these](https://github.com/Nicoshev/windowsTerminal/blob/b180d3a4d1f6a8034eeb19b1a7d262fbfcfabb79/oss/rapidhash/rapidhash.h#L109) into every translation unit? We've got a lot of TUs that include this...

**Comment by Nicoshev** [comment]
Hey @DHowett, thanks a lot for the review!

The global variable should not be replicated accross translation units because rapidhash.h is only used by hash.h, which states '#pragma once' before including it. For extra security, I've added '#pragma once' to rapidhash.h :)

I've also just updated the rapidhash header to a newer version, which has greater compatibility with C++.
The global variable is now marked as 'constexpr' within C++ compilation, hinting the compiler to inline its value whenever used.
Nevertheless, this optimization was still observed when compiling a C file including the header.
The new updated header file also adds the noexcept qualifiers, which allowed the removal of warning supressing code.

Having the 64-bit hash function undocked may be a favorable thing looking forward.
The last commit shows how effortless is to update the hash implementation.
It should be easy to perform any change to the hash function if needed in the future.

wyhash has been deprecated by its owner, it may be counterproductive to use unmaintained code.

**Comment by DHowett** [comment]
Thanks!

`#pragma once` doesn't prevent a header from being included in every translation unit; it just prevents a header from being included more than once in the _same_ TU!

I'm glad that it was easy to update. That is definitely a point in its favor!

FWIW, though, on maintenance: does wyhash _need_ maintenance? Will `rapidhash` need maintenance? I'm firmly of the belief that a project can be "done". If a vulnerability is discovered somehow in the hashing algorithm that would put us at risk, the work is the same to fix it as it is to move to a different hash function. Why move now, from a project that is entirely complete, in the absence of such a vulnerability, for the use cases we have?

**Comment by Nicoshev** [comment]
It is unclear if wyhash or rapidhash will benefit from maintenance.

New language standards may introduce qualifiers or notions from which the hash functions could benefit.

I do agree that this change is not absolutely necesary, although it does leave the codebase modernized and will produce a slight battery life improvement on devices using the terminal.
Additionally, having available the lastest version of the hash function may be useful for a future arising need.

I can also change the PR to implement rapidhash in-place within hash.h, if that's preferable.

Thanks,
Nicolas

**Comment by zadjii-msft** [comment]
It seems like there's not a measurable improvement from changing this at this time. We may as well spare everyone engineering effort by leaving this for now. If in the future there's a measurable improvement to be gained from switching, we can revisit this. Thanks!

**Pull request closed without merging**